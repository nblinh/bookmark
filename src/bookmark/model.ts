import { dataToBookmark } from "../utils/bookmarkUtils";

export class Bookmark {
    view: string;
    url: string;
    title: string;
    author: string;
    addDate: Date;
    publicDate: Date;
    duration?: number; //in seconde
    width?: number;
    height?: number;

    constructor(view: string, url: string, title: string, author: string, publicDate: string) {
        this.view = view;
        this.url = url;
        this.title = title;
        this.author = author;
        this.addDate = new Date();
        this.publicDate = new Date(publicDate);
    }
}

export const getBookmarkInfo = async (url: string): Promise<Bookmark> => {
    if (!url.includes('vimeo.com') && !url.includes('flickr.com')) {
        throw new Error();
    }

    return fetch('https://noembed.com/embed?url=' + url)
        .then(response => response.json())
        .then(data => dataToBookmark(data))
}