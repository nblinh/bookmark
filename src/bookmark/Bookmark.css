.Border {
    border: '1px solid black';
}

.textLeft {
    text-align: left;
    padding-left: 20px;
}

.paddingBottom20 {
    padding-bottom: 20px;
}
  
.Image{
    width: 250px;
    height: 150px;
}

.Video{
    width: 250px;
    height: 150px;
}