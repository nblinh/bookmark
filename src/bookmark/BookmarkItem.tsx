import { FC } from "react";
import { getTimefromDate, secondsToHms } from "../utils/dateUtils";
import DeleteButton from "./DeleteButton";
import { Bookmark } from "./model";

interface BookmarkItemProps {
    bookmarks: Bookmark[]
    setBookmarks: Function;
    bookmark: Bookmark;
}

const BookmarkItem: FC<BookmarkItemProps> = ({ bookmarks, setBookmarks, bookmark }) => {
    return (
        <tr>
            {bookmark.duration && (
                <>
                    <td style={{ border: '1px solid black' }}><iframe title={bookmark.title} src={'https://player.vimeo.com' + bookmark.view.replace('s', '')} width='250' height='150' allow='autoplay; fullscreen; picture-in-picture'></iframe></td>
                    <td style={{ paddingBottom: '20px', border: '1px solid black' }}>
                        <table>
                            <tr><td align='left'><li><b>URL:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.url}</td></tr>
                            <tr><td align='left'><li><b>Titre de la vid√©o:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.title}</td></tr>
                            <tr><td align='left'><li><b>Auteur:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.author}</td></tr>
                            <tr><td align='left'><li><b>Date d'ajout dans l'application:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                                il y a {getTimefromDate(bookmark.addDate)}</td></tr>
                            <tr><td align='left'><li><b>Date de publication sur Vimeo:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                                <span>le {new Intl.DateTimeFormat("fr-FR", { year: "numeric", month: "long", day: "2-digit" }).format(bookmark.publicDate)}</span></td>
                            </tr>
                            <tr><td align='left'><li><b>Duration:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{secondsToHms(bookmark.duration)}</td></tr>
                        </table>
                    </td>
                    <DeleteButton bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmark} />

                </>
            )}
            {!bookmark.duration && (
                <>
                    <td style={{ border: '1px solid black' }}>
                        <img src={bookmark.view} alt='' style={{ width: '250px', height: '150px' }} />
                    </td>
                    <td style={{ border: '1px solid black' }}>
                        <table>
                            <tr><td align='left'><li><b>URL:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.url}</td></tr>
                            <tr><td align='left'><li><b>Titre de la photo:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.title}</td></tr>
                            <tr><td align='left'><li><b>Auteur:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.author}</td></tr>
                            <tr><td align='left'><li><b>Date d'ajout dans l'application:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                                il y a {getTimefromDate(bookmark.addDate)}</td></tr>
                            <tr><td align='left'><li><b>Date de publication sur Flickr:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>
                            </td>
                            </tr>
                            <tr><td align='left'><li><b>Largeur x Hauteur:</b></li></td><td align='left' style={{ paddingLeft: '20px' }}>{bookmark.width + 'x' + bookmark.height}</td></tr>
                        </table>
                    </td>
                    <DeleteButton bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmark} />
                </>
            )}
        </tr>);
}

export default BookmarkItem;