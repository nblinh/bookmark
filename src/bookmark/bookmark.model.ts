export class Bookmark {
    view: string;
    url: string;
    title: string;
    author: string;
    addDate: Date;
    publicDate: Date;
    duration?: number; //in seconde
    width?: number;
    height?: number;

    constructor(view: string, url: string, title: string, author: string, publicDate: string) {
        this.view = view;
        this.url = url;
        this.title = title;
        this.author = author;
        this.addDate = new Date();
        this.publicDate = new Date(publicDate);
    }
}

