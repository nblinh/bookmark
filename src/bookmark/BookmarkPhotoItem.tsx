import { FC } from "react";
import { Bookmark } from "./bookmark.model";
import { getTimefromDate } from "../utils/dateUtils";
import './Bookmark.css'; 

const Border = {
    border: '1px solid black'
};

interface BookmarkPhotoItemProps {
    bookmark: Bookmark;
}
 
const BookmarkPhotoItem: FC<BookmarkPhotoItemProps> = ({bookmark}) => {
    return ( 
        <>
            <td style={Border}>
                <img src={bookmark.view} alt='' className="Image" />
            </td>
            <td style={Border}>
                <table>
                    <tr><td align='left'><li><b>URL:</b></li></td><td className="textLeft">{bookmark.url}</td></tr>
                    <tr><td align='left'><li><b>Titre de la photo:</b></li></td><td className="textLeft">{bookmark.title}</td></tr>
                    <tr><td align='left'><li><b>Auteur:</b></li></td><td className="textLeft">{bookmark.author}</td></tr>
                    <tr><td align='left'><li><b>Date d'ajout dans l'application:</b></li></td><td className="textLeft">
                        {getTimefromDate(bookmark.addDate).trim() !== '' ? 'il y a' : 'Ã  l\'instant'} {getTimefromDate(bookmark.addDate)}</td></tr>
                    <tr><td align='left'><li><b>Date de publication sur Flickr:</b></li></td><td className="textLeft">
                    </td>
                    </tr>
                    <tr><td align='left'><li><b>Largeur x Hauteur:</b></li></td><td className="textLeft">{bookmark.width + 'x' + bookmark.height}</td></tr>
                </table>
            </td>
        </>
     );
}
 
export default BookmarkPhotoItem;