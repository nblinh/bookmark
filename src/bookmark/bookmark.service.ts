import { Bookmark } from "./bookmark.model";
import { dataToBookmark } from "../utils/bookmarkUtils";

export const BookmarkService = {
    getBookmarkInfo: async (url: string): Promise<Bookmark> => {
        if (!url.includes('vimeo.com') && !url.includes('flickr.com')) {
            console.error('Not a valid bookmark url');
            throw new Error('Not a valid bookmark url');
        }
    
        return fetch('https://noembed.com/embed?url=' + url)
            .then(response => response.json())
            .then(data => dataToBookmark(data))
            .catch(error => {
                console.error('There was an error!', error);
                throw new Error('Get bookmark info error');
            });
    }
}