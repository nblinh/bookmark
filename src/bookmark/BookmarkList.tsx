import { FC } from "react";
import BookmarkItem from "./BookmarkItem";
import { Bookmark } from "./model";

interface BookmarkListProps {
    bookmarks: Bookmark[];
    setBookmarks: Function;
}

const BookmarkList: FC<BookmarkListProps> = ({ bookmarks, setBookmarks }) => {
    const showBookmarks = bookmarks.sort((a, b)=>(b.addDate.getTime()-a.addDate.getTime())).map((bookmark: Bookmark) => <BookmarkItem bookmarks={bookmarks} setBookmarks={setBookmarks} bookmark={bookmark} />)
    return (
        <table data-testid='bookmarkTable' style={{ border: '1px solid black', borderCollapse: 'collapse' }}>
            {showBookmarks}
        </table>
    );
}

export default BookmarkList;