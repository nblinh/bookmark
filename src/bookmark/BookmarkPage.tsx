import { FC, useState } from "react";
import BookmarkList from "./BookmarkList";
import { Bookmark, getBookmarkInfo } from "./model";

interface BookmarkPageProps {

}

const BookmarkPage: FC<BookmarkPageProps> = () => {
    const [bookmarks, setBookmarks] = useState<Bookmark[]>([]);
    const [url, setUrl] = useState("");
    const [error, setError] = useState(false);

    const handleSubmit = (evt: any) => {
        evt.preventDefault();

        getBookmarkInfo(url).then(bookmark => {
            bookmarks.push(bookmark);
            setBookmarks([...bookmarks]);
            setError(false);
        })
            .catch((error) => {
                console.error('Error:', error);
                setError(true);
            });
    }
    return (
        <div>
            <form onSubmit={handleSubmit} style={{ paddingBottom: '30px' }}>
                <label>
                    Add URL:
                    <input
                        type="text"
                        value={url}
                        data-testid="urlText"
                        onChange={e => setUrl(e.target.value)}
                    />
                </label>
                <input type="submit" value="Submit" data-testid="submitBtn" />
                <br />
                {error && <span style={{ color: 'red' }}> Please enter a valid URL!This application support only two URL type: flickr.com and vimeo.com</span>}
            </form>

            <BookmarkList bookmarks={bookmarks} setBookmarks={setBookmarks} />
        </div>
    );
}

export default BookmarkPage;