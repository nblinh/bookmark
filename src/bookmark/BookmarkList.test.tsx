
import BookmarkList from "./BookmarkList";
import { render } from '@testing-library/react';
import { Bookmark } from "./bookmark.model";

const bookmarks: Bookmark[] = [
    {
        "view": "/videos/565486457",
        "url": "https://vimeo.com/565486457",
        "title": "Sylvain Lhommée @ Nation Entreprenante - Episode #5",
        "author": "BARTERLINK",
        "addDate": new Date(),
        "publicDate": new Date(),
        "duration": 1070
    }
]

const setBookmarks = jest.fn()

describe('BookmarkList', () => {
    test('should render', () => {
        const { container } = render(<BookmarkList bookmarks={[]} setBookmarks={setBookmarks} />);
        expect(container).toMatchSnapshot();
    });

    test('should have table of bookmark with element', () => {
        const { getByTestId, getByText } = render(<BookmarkList bookmarks={bookmarks} setBookmarks={setBookmarks} />);
        expect(getByTestId('bookmarkTable')).toBeInTheDocument();
        expect(getByText('Sylvain Lhommée @ Nation Entreprenante - Episode #5')).toBeInTheDocument();
    });

});