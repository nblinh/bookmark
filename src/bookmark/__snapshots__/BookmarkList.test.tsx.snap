// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkList should render 1`] = `
<div>
  <table
    data-testid="bookmarkTable"
    style="border: 1px solid black; border-collapse: collapse;"
  />
</div>
`;
